packages/vm/src/context.rs

```diff
@@ -341,33 +348,177 @@ pub fn with_querier_from_context<S: Storage, Q: Querier, F: FnMut(&Q)>(ctx: &Ctx
     b.querier = querier;
 }
 
-pub fn take_storage<S: Storage, Q: Querier>(ctx: &Ctx) -> Option<S> {
+/// take_context_data will return the original storage and querier, and closes any remaining
+/// iterators. This is meant to be called when recycling the instance
+pub(crate) fn take_context_data<S: Storage, Q: Querier>(ctx: &Ctx) -> (Option<S>, Option<Q>) {
     let b = unsafe { get_data::<S, Q>(ctx.data) };
-    // we do this to avoid cleanup
     let mut b = mem::ManuallyDrop::new(b);
-    b.storage.take()
+    // free out the iterator as this finalizes the instance
+    free_iterator(&mut b);
+    (b.storage.take(), b.querier.take())
 }
 
-pub fn leave_storage<S: Storage, Q: Querier>(ctx: &Ctx, storage: Option<S>) {
+/// leave_context_data sets the original storage and querier. These must both be set.
+/// Should be followed by exactly one call to take_context_data when the instance is finished.
+pub(crate) fn leave_context_data<S: Storage, Q: Querier>(ctx: &Ctx, storage: S, querier: Q) {
```

Comment:
Isn't this a `move_into_context_data`, as we're transferring ownership of some elements from the argument into the context? `take_context_data` could then be `move_out_of_context_data`